m-ruslan popup ready for review

m-ruslan opened the pull request:
### HTML & CSS practice: Hooli-style Popup

[Demo](https://m-ruslan.github.io/Kottans-HTML-CSS-practice-Hooli-style-Popup/)
[Code base](https://github.com/m-ruslan/Kottans-HTML-CSS-practice-Hooli-style-Popup)

The code is submitted in a dedicated feature branch.
Please, review.

lenadgit commented:
After clicking on Button "More" the bottom scroll appears. Could you please fix it?

lenadgit commented:
Attribute "alt" is mandatory for each tag "img". Could you please add this attr. for each "img" where it was missed?

lenadgit commented:
According to specification HTML5    http://htmlbook.ru/html/img     the tag "img" must be not closed. Self-closed tags use only in HTML4, XHTML   . Could you fix it please?

Temu4 commented on submissions/m-ruslan/html-css-popup/index.html:
```diff
@@ -0,0 +1,194 @@
+  <body>
+    <nav>
+      <div class="nav-container">
+        <div class="nav-el nav-el-text" tabindex="0">
```
I hope better way to interact with user throw native tags as `a` or `button`.
It will be more clear and [a11y](https://habr.com/ru/post/470091/) code

Temu4 commented on submissions/m-ruslan/html-css-popup/index.html:
```diff
@@ -0,0 +1,194 @@
+          </div>
+<!-- закончил рефакторинг тут -->
+          <section id="apps-window">
+            <div id="top-triangle"></div>
```
Also such "styles elements" better to replace by `css` `:before` or `:after` ([more info](https://developer.mozilla.org/ru/docs/Web/CSS/::before)).

Temu4 commented on submissions/m-ruslan/html-css-popup/index.html:
```diff
@@ -0,0 +1,194 @@
+  </head>
+  <body>
+    <nav>
+      <div class="nav-container">
```
Some unofficial rule of creating header navigation is to use such structure `header>ul>li>a`  ([more info](https://developer.mozilla.org/ru/docs/Web/HTML/Element/nav))

Temu4 commented on submissions/m-ruslan/html-css-popup/index.html:
```diff
@@ -0,0 +1,194 @@
+            
+            <div id="apps-cont">
+              <div id="apps-block1">
+                <div class="app-cont" tabindex="1">
```
Better to use native links vs divs

Temu4 commented on submissions/m-ruslan/html-css-popup/index.html:
```diff
@@ -0,0 +1,194 @@
+                  <p>Calendar</p>
+                </div>
+
+                <div class="app-cont" tabindex="1">
```
You can try to reduce amount of wrappers. For example:
```
<a href="#">
   <img...>
   Mail
</a>

Temu4 reviewed (changes requested):
Well done! :thumb:
Nice visual styling.
But can you take a look at some moments?

MaxKalashnyk reviewed (changes requested):
Good job! Please, check your preview in the [validator](https://validator.w3.org/) and fix all errors there, and we will ready to go ahead.

m-ruslan commented:
@lenadgit, @Temu4, @MaxKalashnyk, thank you very much for the review and feedback.
It seems like I've fixed all the issues in the code you wrote about.
@MaxKalashnyk, I also fixed several errors that were shown by validator from 'https://validator.w3.org/', except two errors show in the screenshot below. As I understand these errors appers due to trying to make pure CSS (without JS) button. Probbably there is a way to avoid this errors, but I haven't found it yet (also I haven't watched code of other students of the course, it seems to me that it's better to do it after my code will be approved)
![image](https://user-images.githubusercontent.com/30605837/97108153-60551a80-16d4-11eb-92e6-c9ee396edac1.png)

MaxKalashnyk commented:
> @lenadgit, @Temu4, @MaxKalashnyk, thank you very much for the review and feedback.
> It seems like I've fixed all the issues in the code you wrote about.
> @MaxKalashnyk, I also fixed several errors that were shown by validator from 'https://validator.w3.org/', except two errors show in the screenshot below. As I understand these errors appears due to trying to make pure CSS (without JS) button. Probably there is a way to avoid this errors, but I haven't found it yet (also I haven't watched code of other students of the course, it seems to me that it's better to do it after my code will be approved)
> ![image](https://user-images.githubusercontent.com/30605837/97108153-60551a80-16d4-11eb-92e6-c9ee396edac1.png)

These two errors still exist because you wrapped your labels to the `a` tag. No need to do it. Just remove `a` tag there and it will work properly

MaxKalashnyk reviewed (changes requested):
Good job! 
Please, reduce the number of ids on your elements. Better to use classes instead of ids.
Also, pay attention to your styles (a lot of your selectors use ids, better to use classes in CSS selectors as well).

One more thing I noticed in your visual preview: when I tried to navigate your page using a keyboard I've noticed that I couldn't jump to the links in your popup. I moved to the next link instead of the links from your popup. Please, fix it as well.

About the other student's works. I think it's ok to see other approved PRs. You can pay your attention to common mistakes that students make and avoid these mistakes in your code.

Temu4 commented on submissions/m-ruslan/html-css-popup/index.html:
```diff
@@ -0,0 +1,153 @@
+
+        <li class="nav-el">
+          <input type="checkbox" id="button-1"  tabindex="0">
+          <a href="#" tabindex="-1">
```
I guess native link tag (`a`) should direct user somewhere, but in this case better to use
`button` and simple `div` tags.

m-ruslan replied:
Thank you, Artem.
I've removed this `a` tag because it caused an error in html validator.

Temu4 commented on submissions/m-ruslan/html-css-popup/index.html:
```diff
@@ -0,0 +1,153 @@
+            
+            <div id="apps-cont">
+              <div id="apps-block1">
+                <a href="#" class="app-cont" tabindex="1">
```
In general (99%) you shouldn't add additional `tabindex` attributes to interactive tags (links, buttons or form elements). 
In this case `tabindex` attribute breaks focusing flow

![Oct-26-2020 12-48-51](https://user-images.githubusercontent.com/42999342/97163663-a4eebd80-1789-11eb-9d64-3573cab2990f.gif)

m-ruslan replied:
Thank you, I've removed tabindexes from my code and it looks that it works fine without them)

Temu4 replied:
It's really better!
But one more thing:
![Oct-30-2020 21-36-24](https://user-images.githubusercontent.com/42999342/97749680-25018400-1af8-11eb-990b-2b5c2d06c622.gif)

You can just move label "under: the links wrapper.
<img width="1012" alt="Screenshot 2020-10-30 at 21 38 21" src="https://user-images.githubusercontent.com/42999342/97749916-86c1ee00-1af8-11eb-8715-678df3ef9769.png">

m-ruslan replied:
Artem, I've moved 'label' as you showed in the screenshot above, but 'More' button still gets focused before the links.
If I move 'input' too, then I can't manage to change '.apps-cont' styles (or haven't found how to do it yet) when 'input' is checked because '.apps-cont' becomes previous sibling of 'input' in the html tree.

Temu4 replied:
Hmm... you're right!
In perfect world, I believe, you have to restructured your element in markup and some styles... but our world isn't perfect 🌏 
You've made a lot of changes and I hope you learned something.

I've marked this issue as "*resolved*"

Temu4 commented on submissions/m-ruslan/html-css-popup/app.css:
```diff
@@ -0,0 +1,248 @@
+  left: -9999px;
+}
+
+.nav-el > a {
```
Better way to use classes for styling. 
Smth like `nav-el__item` (if you use [BEM](https://ru.bem.info/methodology/)).
It will make your code more universal (for example if you change `a` to `button` or `div`).

m-ruslan replied:
Thank you, I've tried to apply basic rules of BEM, please review my updated code.

Temu4 commented on submissions/m-ruslan/html-css-popup/index.html:
```diff
@@ -0,0 +1,153 @@
+          <a href="#" tabindex="0"><img id="bell-icon" src="img/bell.png" alt="Notifications icon"></a>
+        </li>
+        <li class="nav-el">
+          <a href="#" tabindex="0"><img id="profile-photo" src="img/bilbo.jpeg" alt="profile-photo"></a>
```
Try not to load too big images, because it has effect on loading speed.
 
<img width="985" alt="Screenshot 2020-10-26 at 12 29 18" src="https://user-images.githubusercontent.com/42999342/97164385-b5ebfe80-178a-11eb-9ee4-30040eb73c4e.png">

m-ruslan replied:
Thank you, I've reduced the size of profile-photo image.
Also I've noticed that all other images are big sized too, but I havn't changed their size for now due to don't have any installed program that can reduce the size of png files, but I will take this into account in the future.

A-Ostrovnyy commented on submissions/m-ruslan/html-css-popup/index.html:
```diff
@@ -9,185 +9,144 @@
-            <div id="more">
-              <label for="button-2">
-                <div id="more-label-div">More</div>
+            <button id="more" tabindex="-1">
```
You can't nest one interactive element inside another one, don't do it in the future)
All interactive elements(as `<button>, <a>, <input/>, etc.`) by default has focus, no need to use `tabindex` attribute for them, except situation when you want to change your default order of elements.

m-ruslan replied:
Thank you, Alexandr.
I've fixed it in my updated code.

A-Ostrovnyy commented on submissions/m-ruslan/html-css-popup/index.html:
```diff
@@ -9,185 +9,144 @@
-            <div id="top-triangle-border"></div>
+          </a>
 
+          <div id="apps-window">
```
Avoid styling by `#id` and HTLM tag.
Separate responsibilities:
- HTML tag for markup;
- CSS class for styling;
- ID for JS logic

m-ruslan replied:
Thank you for the explanation, haven't thought/read about such separate responsibilities before.

A-Ostrovnyy reviewed (changes requested):
Good work!
Fix order of focused elements: if popup is open then focus from menu item must go into popup.
Don't hesitate to ask if you don't understand something or you need help.
See my comment above.

m-ruslan commented:
@MaxKalashnyk, @Temu4, @A-Ostrovnyy thank you for one more stage of review.
I've tried to fix issues you wrote about, please review one more time)

m-ruslan commented:
> These two errors still exist because you wrapped your labels into the `a` tag. No need to do it. Just remove `a` tag there and it will work properly

@MaxKalashnyk, thank you, i've removed `a` tag and `button` tag that wrapped the `label` tag and it works without errors now.

m-ruslan commented:
> Good job!
> Please, reduce the number of ids on your elements. Better to use classes instead of ids.
> Also, pay attention to your styles (a lot of your selectors use ids, better to use classes in CSS selectors as well).
> 
> One more thing I noticed in your visual preview: when I tried to navigate your page using a keyboard I've noticed that I couldn't jump to the links in your popup. I moved to the next link instead of the links from your popup. Please, fix it as well.
> 
> About the other student's works. I think it's ok to see other approved PRs. You can pay your attention to common mistakes that students make and avoid these mistakes in your code.

@MaxKalashnyk,
- I've replaced ids with classes in html and css.
- It seems that now navigation works better.
- Thank you for the advice, It's a great to look at reviews in other PRs to understand common mistakes.

MaxKalashnyk reviewed (approved):
Good job!

A-Ostrovnyy commented on submissions/m-ruslan/html-css-popup/index.html:
```diff
@@ -4,134 +4,130 @@
                   
-                  <a href="#" class="app-cont" tabindex="1"> 
-                    <img src="img/chat.png"  alt="Chat icon">
+                  <a href="#" class="app-block__item"> 
```
Your parent block is `appS-block` and child elements is `app-block__item`. Is it right?

m-ruslan replied:
Thank you.
I've fixed it)

A-Ostrovnyy reviewed (changes requested):
Good work!
One small issue)

Temu4 commented on submissions/m-ruslan/html-css-popup/index.html:
```diff
@@ -0,0 +1,149 @@
+            <div class="apps-cont">
+              <div class="apps-block">
+                <a href="#" class="app-block__item">
+                  <img src="img/calendar.png" class="app-block__img" alt="Calendar icon">
```
Can you please check size of the images?
<img width="825" alt="Screenshot 2020-10-30 at 21 42 34" src="https://user-images.githubusercontent.com/42999342/97750342-444ce100-1af9-11eb-8f64-0904dddf2d40.png">

m-ruslan replied:
Thank you.
I've fixed it)

m-ruslan commented:
> Good job!

@MaxKalashnyk thank you)

OleksiyRudenko commented:
@Temu4 @A-Ostrovnyy are we good with this PR?

A-Ostrovnyy reviewed (approved):
LGTM
Good  work

m-ruslan commented:
@A-Ostrovnyy thank you)

@OleksiyRudenko thank you for reminding)

OleksiyRudenko commented:
@m-ruslan once you're done with comments of a reviewer re-request review.
![image](https://user-images.githubusercontent.com/8686631/97809372-27d5b380-1c75-11eb-9800-bcb83f4ae7c7.png)
Reviewers need to get notified as hardly they monitor contsantly anything has changed to better in multiple PRs they're overseeing.

We steel need to make @Temu4 happy ;)

m-ruslan commented:
@OleksiyRudenko thank you, I will use a "re-request review" in the future.

Temu4 reviewed (approved):
Looks good to me!

m-ruslan commented:
> Looks good to me!

@Temu4 I've restructured elements and change stiles a little bit and managed to set right focus flow)
Could you please check my [demo](https://m-ruslan.github.io/Kottans-HTML-CSS-practice-Hooli-style-Popup/)

Temu4 commented:
@m-ruslan now it works perfect!
Well done! 👍

m-ruslan commented:
@Temu4 thank you)
